File: apps/ios/UNES/Features/FolioRunner/FolioSprite.swift
import SwiftUI

/// Folio — the paper-corner mascot from `mascot-concepts.jsx` (Concept 02),
/// translated from SVG into a Canvas-friendly drawer. Five poses cover a
/// Chrome-dino style runner: idle, runA/runB, jump, duck.
enum FolioPose {
    case idle, runA, runB, jump, duck
}

/// Static sprite drawer. Authored on a 200×200 viewBox; the caller passes
/// a destination `frame` and the drawer scales into it. Used both by the
/// game's main Canvas (which packs the whole scene into one renderer) and
/// by `FolioSpriteView` for static placements.
enum FolioSprite {
    static func draw(
        in ctx: GraphicsContext,
        pose: FolioPose,
        frame: CGRect,
        blink: Bool = false
    ) {
        let cfg = config(for: pose)
        let isJump = pose == .jump
        let isDuck = pose == .duck
        // Eyes are drawn closed in duck (matches the squash) or whenever
        // the caller asks for a blink — the game uses duck only, the
        // empty-state mascot uses blink.
        let eyesClosed = isDuck || blink

        var sprite = ctx
        sprite.translateBy(x: frame.minX, y: frame.minY)
        sprite.scaleBy(x: frame.width / 200, y: frame.height / 200)

        // Shadow on the ground line — shrinks while airborne.
        let shadowCY: CGFloat = isJump ? 178 : 174
        let shadowRX: CGFloat = isJump ? 22 : 38
        sprite.fill(
            Path(ellipseIn: CGRect(
                x: 100 - shadowRX, y: shadowCY - 4,
                width: shadowRX * 2, height: 8
            )),
            with: .color(.black.opacity(0.14))
        )

        // Body transform: translate, rotate, squash. Mirrors the SVG's
        // nested <g transform> so the dog-ear and rule lines move with the
        // page.
        var body = sprite
        body.translateBy(x: 100, y: 100 + cfg.bodyY)
        body.rotate(by: .degrees(cfg.bodyR))
        body.scaleBy(x: 1, y: cfg.squash)
        body.translateBy(x: -100, y: -100)

        let bottomBase: CGFloat = isDuck ? 130 : 145
        let bottomMid: CGFloat = isDuck ? 134 : 149
        let bottomCurve: CGFloat = isDuck ? 138 : 153
        let paper = Path { p in
            p.move(to: CGPoint(x: 50, y: 50))
            p.addLine(to: CGPoint(x: 138, y: 50))
            p.addLine(to: CGPoint(x: 152, y: 64))
            p.addLine(to: CGPoint(x: 152, y: bottomBase))
            p.addQuadCurve(
                to: CGPoint(x: 100, y: bottomMid),
                control: CGPoint(x: 130, y: bottomCurve)
            )
            p.addQuadCurve(
                to: CGPoint(x: 50, y: bottomMid),
                control: CGPoint(x: 70, y: bottomCurve)
            )
            p.closeSubpath()
        }
        body.fill(paper, with: .color(cream))
        body.stroke(
            paper, with: .color(ink),
            style: StrokeStyle(lineWidth: 2.5, lineJoin: .round)
        )

        // Folded corner (dog ear) — pivots independently around its own
        // hinge so the run cycle wags it.
        var ear = body
        ear.translateBy(x: 145, y: 57)
        ear.rotate(by: .degrees(cfg.earR))
        ear.translateBy(x: -145, y: -57)
        let earPath = Path { p in
            p.move(to: CGPoint(x: 138, y: 50))
            p.addLine(to: CGPoint(x: 152, y: 64))
            p.addLine(to: CGPoint(x: 138, y: 64))
            p.closeSubpath()
        }
        ear.fill(earPath, with: .color(creamDark))
        ear.stroke(earPath, with: .color(ink), style: StrokeStyle(lineWidth: 2))

        // Coral margin rule.
        let marginEnd: CGFloat = isDuck ? 132 : 144
        var margin = Path()
        margin.move(to: CGPoint(x: 68, y: 58))
        margin.addLine(to: CGPoint(x: 68, y: marginEnd))
        body.stroke(
            margin, with: .color(coral),
            style: StrokeStyle(lineWidth: 2, lineCap: .round)
        )

        // Faint horizontal rule lines.
        for y in [78.0, 92.0, 106.0] as [CGFloat] {
            var line = Path()
            line.move(to: CGPoint(x: 78, y: y))
            line.addLine(to: CGPoint(x: 148, y: y))
            body.stroke(
                line, with: .color(ink.opacity(0.10)),
                style: StrokeStyle(lineWidth: 1)
            )
        }

        if !eyesClosed {
            body.fill(
                Path(ellipseIn: CGRect(x: 98 - 3.5, y: 86 - 3.5, width: 7, height: 7)),
                with: .color(ink)
            )
            body.fill(
                Path(ellipseIn: CGRect(x: 120 - 3.5, y: 86 - 3.5, width: 7, height: 7)),
                with: .color(ink)
            )
            body.fill(
                Path(ellipseIn: CGRect(x: 99 - 1, y: 85 - 1, width: 2, height: 2)),
                with: .color(cream)
            )
            body.fill(
                Path(ellipseIn: CGRect(x: 121 - 1, y: 85 - 1, width: 2, height: 2)),
                with: .color(cream)
            )
        } else {
            // Sit the lids at the open-eye Y in non-duck poses so a blink
            // looks like eyes shutting in place; duck keeps its lower
            // squashed-head position.
            let lidY: CGFloat = isDuck ? 92 : 86
            var leftLid = Path()
            leftLid.move(to: CGPoint(x: 94, y: lidY))
            leftLid.addLine(to: CGPoint(x: 102, y: lidY))
            body.stroke(
                leftLid, with: .color(ink),
                style: StrokeStyle(lineWidth: 2.5, lineCap: .round)
            )
            var rightLid = Path()
            rightLid.move(to: CGPoint(x: 116, y: lidY))
            rightLid.addLine(to: CGPoint(x: 124, y: lidY))
            body.stroke(
                rightLid, with: .color(ink),
                style: StrokeStyle(lineWidth: 2.5, lineCap: .round)
            )
        }

        var smile = Path()
        smile.move(to: CGPoint(x: 102, y: 100))
        smile.addQuadCurve(to: CGPoint(x: 116, y: 100), control: CGPoint(x: 109, y: 104))
        body.stroke(
            smile, with: .color(ink),
            style: StrokeStyle(lineWidth: 2, lineCap: .round)
        )

        // Rippled bottom edge as legs — drawn on the parent transform (not
        // the body's) so the squash doesn't deform the gait.
        if !isJump && !isDuck {
            var legs = sprite
            legs.translateBy(x: 0, y: cfg.bodyY * 0.6)
            var legPath = Path()
            legPath.move(to: CGPoint(x: 70, y: 152))
            legPath.addQuadCurve(
                to: CGPoint(x: 86, y: 152),
                control: CGPoint(x: 78, y: 162 + cfg.rippleA)
            )
            legPath.addQuadCurve(
                to: CGPoint(x: 102, y: 152),
                control: CGPoint(x: 94, y: 162 - cfg.rippleA)
            )
            legPath.addQuadCurve(
                to: CGPoint(x: 118, y: 152),
                control: CGPoint(x: 110, y: 162 + cfg.rippleA)
            )
            legPath.addQuadCurve(
                to: CGPoint(x: 134, y: 152),
                control: CGPoint(x: 126, y: 162 - cfg.rippleA)
            )
            legs.stroke(
                legPath, with: .color(ink),
                style: StrokeStyle(lineWidth: 2.5, lineCap: .round)
            )
        }
    }

    private struct Config {
        var bodyY: CGFloat
        var bodyR: CGFloat
        var rippleA: CGFloat
        var squash: CGFloat
        var earR: CGFloat
    }

    private static func config(for pose: FolioPose) -> Config {
        switch pose {
        case .idle: return Config(bodyY: 0,   bodyR: 0,  rippleA: 0,  squash: 1.0,  earR: 0)
        case .runA: return Config(bodyY: -2,  bodyR: -2, rippleA: 6,  squash: 1.02, earR: -3)
        case .runB: return Config(bodyY: -2,  bodyR: 2,  rippleA: -6, squash: 1.02, earR: 3)
        case .jump: return Config(bodyY: -22, bodyR: -6, rippleA: 0,  squash: 1.06, earR: -8)
        case .duck: return Config(bodyY: 12,  bodyR: 0,  rippleA: 0,  squash: 0.6,  earR: 0)
        }
    }

    // Fixed palette from `mascot-concepts.jsx`. Not adaptive — the runner
    // is a paper-themed scene and reads the same in light/dark.
    static let cream     = Color(red: 0xFB / 255, green: 0xF7 / 255, blue: 0xF2 / 255)
    static let creamDark = Color(red: 0xE9 / 255, green: 0xE0 / 255, blue: 0xD2 / 255)
    static let ink       = Color(red: 0x1A / 255, green: 0x14 / 255, blue: 0x20 / 255)
    static let coral     = Color(red: 0xE8 / 255, green: 0x5D / 255, blue: 0x4E / 255)
    static let plum      = Color(red: 0x2D / 255, green: 0x1B / 255, blue: 0x4E / 255)
    static let amber     = Color(red: 0xF4 / 255, green: 0xA2 / 255, blue: 0x3C / 255)
    static let peach     = Color(red: 0xFB / 255, green: 0xD9 / 255, blue: 0xA8 / 255)
}

/// Static SwiftUI wrapper — used outside the running game's Canvas (e.g.
/// the start overlay).
struct FolioSpriteView: View {
    var pose: FolioPose
    var size: CGFloat = 90
    var blink: Bool = false

    var body: some View {
        Canvas { ctx, _ in
            FolioSprite.draw(
                in: ctx,
                pose: pose,
                frame: CGRect(x: 0, y: 0, width: size, height: size),
                blink: blink
            )
        }
        .frame(width: size, height: size)
    }
}

/// Idle Folio that closes its eyes once or twice every ~25 seconds.
/// Used as the empty-state mascot in `DayColumn` — also acts as the
/// long-press target for the runner easter egg.
struct BlinkingFolio: View {
    var size: CGFloat = 96

    @State private var blink = false

    var body: some View {
        FolioSpriteView(pose: .idle, size: size, blink: blink)
            .task { await runBlinkLoop() }
    }

    private func runBlinkLoop() async {
        while !Task.isCancelled {
            // Random gap (avg ~25s) between blink events. Re-rolled each
            // loop so the rhythm doesn't feel mechanical.
            try? await Task.sleep(for: .seconds(.random(in: 18 ... 32)))
            await singleBlink()
            // Roughly a third of the time, follow up with a quick double
            // blink — feels natural without becoming twitchy.
            if Int.random(in: 0 ..< 3) == 0 {
                try? await Task.sleep(for: .milliseconds(110))
                await singleBlink()
            }
        }
    }

    private func singleBlink() async {
        blink = true
        try? await Task.sleep(for: .milliseconds(140))
        blink = false
    }
}

File: apps/android/design-system/src/main/kotlin/dev/forcetower/unes/designsystem/components/FolioSprite.kt
package dev.forcetower.unes.designsystem.components

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.drawscope.withTransform
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import kotlinx.coroutines.delay
import kotlin.random.Random

// Folio — the paper-corner mascot from the iOS `FolioSprite` (Concept 02 in
// `mascot-concepts.jsx`), translated into a Compose Canvas drawer. Five poses
// cover the Chrome-dino style runner: Idle, RunA/RunB, Jump, Duck. The drawing
// is authored on a 200×200 viewBox; callers pass a destination rect and the
// drawer scales into it.
enum class FolioPose { Idle, RunA, RunB, Jump, Duck }

// Fixed paper-themed palette, kept in sync with iOS `FolioSprite`. Not adaptive
// — the runner is its own paper-cutout world that should read identically in
// light and dark, so we expose it as a dedicated palette rather than routing
// through `MaterialTheme.melon.brand` (whose hexes don't match exactly).
object FolioPalette {
    val cream = Color(0xFFFBF7F2)
    val creamDark = Color(0xFFE9E0D2)
    val ink = Color(0xFF1A1420)
    val coral = Color(0xFFE85D4E)
    val plum = Color(0xFF2D1B4E)
    val amber = Color(0xFFF4A23C)
    val peach = Color(0xFFFBD9A8)
}

private data class FolioConfig(
    val bodyY: Float,
    val bodyR: Float,
    val rippleA: Float,
    val squash: Float,
    val earR: Float,
)

private fun configFor(pose: FolioPose): FolioConfig = when (pose) {
    FolioPose.Idle -> FolioConfig(0f, 0f, 0f, 1.0f, 0f)
    FolioPose.RunA -> FolioConfig(-2f, -2f, 6f, 1.02f, -3f)
    FolioPose.RunB -> FolioConfig(-2f, 2f, -6f, 1.02f, 3f)
    FolioPose.Jump -> FolioConfig(-22f, -6f, 0f, 1.06f, -8f)
    FolioPose.Duck -> FolioConfig(12f, 0f, 0f, 0.6f, 0f)
}

// Canvas drawer used both by the runner's main scene canvas and by the static
// `FolioSpriteView`. `frame` is in the parent canvas's coordinate space.
fun DrawScope.drawFolio(
    pose: FolioPose,
    frame: Rect,
    blink: Boolean = false,
) {
    val cfg = configFor(pose)
    val isJump = pose == FolioPose.Jump
    val isDuck = pose == FolioPose.Duck
    // Eyes are drawn closed for duck (matches the squash) and whenever the
    // caller asks for a blink — duck is used by the game, blink by the idle
    // empty-state mascot.
    val eyesClosed = isDuck || blink

    withTransform({
        translate(frame.left, frame.top)
        scale(frame.width / 200f, frame.height / 200f, pivot = Offset.Zero)
    }) {
        // Shadow on the ground line — shrinks while airborne.
        val shadowCY = if (isJump) 178f else 174f
        val shadowRX = if (isJump) 22f else 38f
        drawOval(
            color = Color.Black.copy(alpha = 0.14f),
            topLeft = Offset(100f - shadowRX, shadowCY - 4f),
            size = Size(shadowRX * 2f, 8f),
        )

        // Body transform — translate, rotate around (100, 100+bodyY), squash.
        // Mirrors the SVG's nested <g transform> so the dog-ear and rule lines
        // move with the page.
        withTransform({
            translate(100f, 100f + cfg.bodyY)
            rotate(cfg.bodyR, pivot = Offset.Zero)
            scale(1f, cfg.squash, pivot = Offset.Zero)
            translate(-100f, -100f)
        }) {
            val bottomBase = if (isDuck) 130f else 145f
            val bottomMid = if (isDuck) 134f else 149f
            val bottomCurve = if (isDuck) 138f else 153f
            val paper = Path().apply {
                moveTo(50f, 50f)
                lineTo(138f, 50f)
                lineTo(152f, 64f)
                lineTo(152f, bottomBase)
                quadraticTo(130f, bottomCurve, 100f, bottomMid)
                quadraticTo(70f, bottomCurve, 50f, bottomMid)
                close()
            }
            drawPath(paper, color = FolioPalette.cream)
            drawPath(
                paper,
                color = FolioPalette.ink,
                style = Stroke(width = 2.5f, join = StrokeJoin.Round),
            )

            // Folded corner (dog ear) — pivots around its own hinge so the run
            // cycle wags it independently of the body squash.
            withTransform({
                translate(145f, 57f)
                rotate(cfg.earR, pivot = Offset.Zero)
                translate(-145f, -57f)
            }) {
                val earPath = Path().apply {
                    moveTo(138f, 50f)
                    lineTo(152f, 64f)
                    lineTo(138f, 64f)
                    close()
                }
                drawPath(earPath, color = FolioPalette.creamDark)
                drawPath(
                    earPath,
                    color = FolioPalette.ink,
                    style = Stroke(width = 2f),
                )
            }

            // Coral margin rule.
            val marginEnd = if (isDuck) 132f else 144f
            val margin = Path().apply {
                moveTo(68f, 58f)
                lineTo(68f, marginEnd)
            }
            drawPath(
                margin,
                color = FolioPalette.coral,
                style = Stroke(width = 2f, cap = StrokeCap.Round),
            )

            // Faint horizontal rule lines.
            for (y in listOf(78f, 92f, 106f)) {
                val line = Path().apply {
                    moveTo(78f, y)
                    lineTo(148f, y)
                }
                drawPath(
                    line,
                    color = FolioPalette.ink.copy(alpha = 0.10f),
                    style = Stroke(width = 1f),
                )
            }

            if (!eyesClosed) {
                drawOval(
                    color = FolioPalette.ink,
                    topLeft = Offset(98f - 3.5f, 86f - 3.5f),
                    size = Size(7f, 7f),
                )
                drawOval(
                    color = FolioPalette.ink,
                    topLeft = Offset(120f - 3.5f, 86f - 3.5f),
                    size = Size(7f, 7f),
                )
                drawOval(
                    color = FolioPalette.cream,
                    topLeft = Offset(99f - 1f, 85f - 1f),
                    size = Size(2f, 2f),
                )
                drawOval(
                    color = FolioPalette.cream,
                    topLeft = Offset(121f - 1f, 85f - 1f),
                    size = Size(2f, 2f),
                )
            } else {
                // Sit lids at the open-eye Y in non-duck poses so a blink looks
                // like eyes shutting in place; duck keeps the squashed-head
                // lower position.
                val lidY = if (isDuck) 92f else 86f
                val leftLid = Path().apply {
                    moveTo(94f, lidY)
                    lineTo(102f, lidY)
                }
                drawPath(
                    leftLid,
                    color = FolioPalette.ink,
                    style = Stroke(width = 2.5f, cap = StrokeCap.Round),
                )
                val rightLid = Path().apply {
                    moveTo(116f, lidY)
                    lineTo(124f, lidY)
                }
                drawPath(
                    rightLid,
                    color = FolioPalette.ink,
                    style = Stroke(width = 2.5f, cap = StrokeCap.Round),
                )
            }

            val smile = Path().apply {
                moveTo(102f, 100f)
                quadraticTo(109f, 104f, 116f, 100f)
            }
            drawPath(
                smile,
                color = FolioPalette.ink,
                style = Stroke(width = 2f, cap = StrokeCap.Round),
            )
        }

        // Rippled bottom edge as legs — drawn on the parent (sprite) transform,
        // not the body's, so the squash doesn't deform the gait.
        if (!isJump && !isDuck) {
            withTransform({
                translate(0f, cfg.bodyY * 0.6f)
            }) {
                val legPath = Path().apply {
                    moveTo(70f, 152f)
                    quadraticTo(78f, 162f + cfg.rippleA, 86f, 152f)
                    quadraticTo(94f, 162f - cfg.rippleA, 102f, 152f)
                    quadraticTo(110f, 162f + cfg.rippleA, 118f, 152f)
                    quadraticTo(126f, 162f - cfg.rippleA, 134f, 152f)
                }
                drawPath(
                    legPath,
                    color = FolioPalette.ink,
                    style = Stroke(width = 2.5f, cap = StrokeCap.Round),
                )
            }
        }
    }
}

// Static SwiftUI-style wrapper — used outside the running game's Canvas (e.g.
// the runner's start overlay and the schedule empty-state mascot).
@Composable
fun FolioSpriteView(
    pose: FolioPose,
    modifier: Modifier = Modifier,
    size: Dp = 90.dp,
    blink: Boolean = false,
) {
    Canvas(modifier = modifier.size(size)) {
        drawFolio(
            pose = pose,
            frame = Rect(0f, 0f, this.size.width, this.size.height),
            blink = blink,
        )
    }
}

// Idle Folio that closes its eyes once or twice every ~25 seconds. Used as the
// schedule empty-state mascot and as the long-press target for the runner
// easter egg.
@Composable
fun BlinkingFolio(
    modifier: Modifier = Modifier,
    size: Dp = 96.dp,
) {
    var blink by remember { mutableStateOf(false) }
    LaunchedEffect(Unit) {
        while (true) {
            // Random gap (avg ~25s) between blink events. Re-rolled each loop
            // so the rhythm doesn't feel mechanical.
            delay(Random.nextLong(18_000L, 32_000L))
            blink = true
            delay(140L)
            blink = false
            // Roughly a third of the time, follow up with a quick double blink
            // — feels natural without becoming twitchy.
            if (Random.nextInt(3) == 0) {
                delay(110L)
                blink = true
                delay(140L)
                blink = false
            }
        }
    }
    FolioSpriteView(pose = FolioPose.Idle, size = size, blink = blink, modifier = modifier)
}
